File: kafka-connect-common/src/main/scala/com/datamountaineer/streamreactor/connect/config/PayloadFields.scala
package com.datamountaineer.streamreactor.connect.config

import io.confluent.common.config.ConfigException


/**
  * Contains the SinkConnect payload fields to consider and/or their mappings
  *
  * @param includeAllFields
  * @param fieldsMappings
  */
case class PayloadFields(includeAllFields: Boolean,
                         fieldsMappings: Map[String, String])


object PayloadFields {
  /**
    * Works out the fields and their mappings to be used when inserting a new Hbase row
    *
    * @param setting - The configuration specifing the fields and their mappings
    * @return A dictionary of fields and their mappings alongside a flag specifying if all fields should be used. If no mapping has been specified the field name is considered to be the mapping
    */
  def apply(setting: Option[String]): PayloadFields = {
    setting match {
      case None => PayloadFields(true, Map.empty[String, String])
      case Some(c) =>

        val mappings = c.split(",").map { case f =>
          f.trim.split("=").toSeq match {
            case Seq(field) =>
              field -> field
            case Seq(field, alias) =>
              field -> alias
            case _ => throw new ConfigException(s"$c is not valid. Need to set the fields and mappings like: field1,field2,field3=alias3,[field4, field5=alias5]")
          }
        }.toMap

        PayloadFields(mappings.contains("*"), mappings - "*")
    }
  }
}

File: kafka-connect-common/src/main/java/com/datamountaineer/streamreactor/connect/config/PayloadFields.java
/**
 * Copyright 2015 Datamountaineer.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 **/

package com.datamountaineer.streamreactor.connect.config;

import com.google.common.base.Joiner;
import io.confluent.common.config.ConfigException;

import java.util.HashMap;
import java.util.Map;

/**
 * Contains the SinkConnect payload fields to consider and/or their mappings
 */
public final class PayloadFields {

    private final Boolean includeAllFields;
    private final Map<String, String> fieldsMappings;

    public PayloadFields(Boolean includeAllFields, Map<String, String> fieldsMappings) {
        this.includeAllFields = includeAllFields;
        this.fieldsMappings = fieldsMappings;
    }

    public PayloadFields() {
        this(true, new HashMap<String, String>());
    }

    public Boolean getIncludeAllFields() {
        return includeAllFields;
    }

    public Map<String, String> getFieldsMappings() {
        return fieldsMappings;
    }

    public static PayloadFields from(String value) {
        if (value == null)
            return new PayloadFields();

        final Map<String, String> fieldAlias = new HashMap<>();
        for (String split : value.split(",")) {
            final String[] arr = split.trim().split("=");
            if (arr.length == 1) {
                fieldAlias.put(arr[0], arr[0]);
            } else if (arr.length == 2) {
                fieldAlias.put(arr[0], arr[1]);
            } else
                throw new ConfigException(value + " is not valid. Need to set the fields and mappings like: field1,field2,field3=alias3,[field4, field5=alias5]");
        }

        final Boolean allFields = fieldAlias.remove("*") != null;

        return new PayloadFields(allFields, fieldAlias);
    }

    @Override
    public String toString() {
        Joiner.MapJoiner mapJoiner = Joiner.on(',').withKeyValueSeparator("=");
        return String.format("PayloadFields(%s,%s)", includeAllFields.toString(), mapJoiner.join(fieldsMappings));
    }
}
